experimental/Geometry/Multiindices.jl

```diff
@@ -0,0 +1,221 @@
+  i = OrderedMultiindex( n, p )
+  if l > binomial( n, p ) 
+    return nothing
+  end
+  for k in 1:p
+    while l > binomial( n-i.index[k], p-k )
+      l = l - binomial( n-i.index[k], p-k )
+      i.index[k] = i.index[k]+1
+    end
+    for r = k+1:p
+      i.index[r] = i.index[r-1]+1
+    end
+  end
+  return i
+end
+
+function index_signature( alpha::OrderedMultiindex )
+  sign = 1
+  for k in alpha.p:-1:1
+    if mod( k-alpha.index[k], 2 ) != 0
+      sign = sign*(-1)
+    end
+  end
+  return sign
+end
+
+mutable struct HomogMultiindex
+  n::Int		# The number of variables
+  d::Int		# The degree
+  a::Vector{Int}	# The multi-exponent
+  i::OrderedMultiindex	# The underlying multiindex
+
+
+  function HomogMultiindex( n::Int, d::Int, a::Vector{Int}, i::OrderedMultiindex )
+    if n <= 0 
+      error( "invalid number of variables" )
+    end
+    if d < 0
+      error( "invalid degree." )
+    end
+    if sum(a) != d 
+      error( "values of a don't match the requested degree." )
+    end 
+    return new( n, d, a, i )
+  end
+  
+  function HomogMultiindex( n::Int, d::Int )
+    if n<= 0
+      error( "invalid number of variables." )
+    end
+    if d<0 
+      error( "invalid degree." )
+    end
+    a = vcat( [ 0 for i in (2:n) ], [d] )
+    i = OrderedMultiindex( d+n-1, n-1 ) 
+    #@show i
+    #@show i.index
+    #@show i.n
+    return HomogMultiindex( n, d, a, i )
+  end
+end
+
+function extract_exponent( alpha::HomogMultiindex )
+  #@show alpha.i.index, alpha.i.p
+  if alpha.i.p > 0
+    a = [ alpha.i.index[1] - 1 ]
+    a = vcat( a, [ alpha.i.index[k+1] - alpha.i.index[k] - 1 for k in (1:alpha.i.p-1) ] )
+    a = vcat( a, [ alpha.i.n - alpha.i.index[alpha.i.p] ])
+  else
+    a = [ alpha.d ]
+  end
+  return a
+end
+
+function Base.show( io::Base.TTY, alpha::HomogMultiindex ) 
+  outstr = "( "
+  for k in (1:alpha.n-1) 
+    outstr = outstr * string(alpha.a[k]) * ", "
+  end
+  #@show alpha.a
+  #@show alpha.d
+  outstr = outstr * "$(alpha.a[alpha.n]) )\n"
+  Base.print( io, outstr )
+end
+  
+function Base.iterate( alpha::HomogMultiindex )
+  Base.iterate( alpha.i )
+  alpha.a = extract_exponent( alpha )
+  return alpha, alpha.a 
+end
+
+function Base.iterate( alpha::HomogMultiindex, state )
+  #@show alpha.i
+  x = Base.iterate( alpha.i, state )
+  if x == nothing 
+    return nothing	# Iteration is exhausted
+  end
+  #@show x
+  #@show typeof( x )
+  alpha.i = x[1]
+  alpha.a = extract_exponent( alpha )
+  return alpha, alpha.a
+end
+
+function power( x::Vector, alpha::HomogMultiindex )
+  n = length( x )
+  if n != alpha.n
+    error( "number of variables not compatible." )
+  end
+  result = 1
+  for k in (1:n)
+    result = result * x[k]^(alpha.a[k])
+  end
+  return result
+end
+
+function linear_index( alpha::HomogMultiindex ) 
+  return linear_index( alpha.i )
+end
+
+# Use Ordered multiindices for indexing 2-dimensional arrays
+
+function Base.getindex( v::Array{Any,2}, α::OrderedMultiindex )
+  if α.p < 2 
```

Comment:
Yes please, I do not want to spend my day googling how to type alpha (or to start learning unicodes by heart).

Reply:
Same with `Γ` below.

Reply:
But this is a very nice feature of Julia! 

It works like this: You type the latex command, say \Gamma, and then you hit the TAB-key. Simsalabim: There comes your unicode character! 

In this particular place, it can be avoided, but in general I think it can significantly increase readability since it's much closer to the mathematical way of writing things.

Reply:
But I don't use Julia to type in my code. And when I need to edit this Julia file, I'll now have to deal with those special letters, and in most editors, they are *not* convenient to enter.

Reply:
Then, I think, one should use a proper editor... like Vim with the Julia plugin.

No, jokes aside. I see that there are reasons to keep the actual code in ASCII letters. But for explanatory comments I will keep using the unicode feature. In the end, the use of such letters is intended by design of the language and in my opinion, people should not blame this on their editors, but adjust their working environment to the language that they would like to work with.

Reply:
We are trying to find/found a minimal common denominator to work with in this project and this is not helpful in this regard.

Reply:
Ok, I see. Even in the comments? I saw this in the code of the intersection theory module and I was delighted how nice everything looked.

Reply:
I don't mind it in the comments. But be prepared that it might go away once someone has to update the comments to reflect changes in the code. Not sure what @fingolfin thinks about unicode in comments.